search/mongo_search.go

```diff
@@ -359,8 +329,136 @@ func (m *MongoSearcher) createParamObjects(params []SearchParam) []bson.M {
 }
 
 func (m *MongoSearcher) createPipelineObject(query Query) []bson.M {
-	// Currently this is just a $match stage with the same query that would be used in find()
-	return []bson.M{{"$match": m.createQueryObject(query)}}
+	standardSearchParams := []SearchParam{}
+	chainedSearchParams := []SearchParam{}
+
+	// Separate out chained search parameters
+	for _, p := range query.Params() {
+		if usesChainedSearch(p) {
+			chainedSearchParams = append(chainedSearchParams, p)
+			continue
+		}
+		standardSearchParams = append(standardSearchParams, p)
+	}
+
+	// Process standard SearchParams
+	pipeline := []bson.M{{"$match": m.createQueryObjectFromParams(standardSearchParams)}}
+
+	// Process chained search parameters
+	for _, p := range chainedSearchParams {
+		pipeline = append(pipeline, m.createChainedSearchPipelineStages(p)...)
+	}
+
+	return pipeline
+}
+
+// The SearchParam argument should be either a ReferenceParam or an OrParam.
+func (m *MongoSearcher) createChainedSearchPipelineStages(searchParam SearchParam) []bson.M {
+	// This returns 2 stages in the pipeline that represent a chained query reference:
+	// 1. A $lookup for the foreign Resource being referenced
+	// 2. A $match on that foreign Resource
+	stages := make([]bson.M, 2)
+
+	// Build the $lookup. We need to get a ReferenceParam (of type ChainedQueryReference)
+	// that we can use to populate the $lookup. If it's an OR, any one of its Items
+	// should do.
+	var lookupRef *ReferenceParam
+	var ok bool
+
+	_, isOr := searchParam.(*OrParam)
+
+	if isOr {
+		lookupRef, ok = searchParam.(*OrParam).Items[0].(*ReferenceParam)
+		if !ok {
+			panic(createInternalServerError("", ""))
+		}
+	} else {
+		lookupRef = searchParam.(*ReferenceParam)
+	}
+
+	chainedRef, ok := lookupRef.Reference.(ChainedQueryReference)
+	if !ok {
+		panic(createInternalServerError("", ""))
+	}
+
+	collectionName := models.PluralizeLowerResourceName(chainedRef.Type)
+
+	stages[0] = bson.M{"$lookup": bson.M{
+		"from":         collectionName,
+		"localField":   convertSearchPathToMongoField(lookupRef.getInfo().Paths[0].Path) + ".referenceid",
```

Comment:
This is only handling the first path.  If there are multiple paths associated with the reference param, then you will miss all the matches for subsequent paths.  To do this right, you need to do a lookup for each path (putting each set of results in its own "as" field).  Then when you prepend the "as" to the paths, you need to keep track of which path goes to which "as".  It's kind of complicated, but not terribly so -- and is really necessary for a correct implementation.  We can discuss if you'd like...